segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -174,77 +185,77 @@ public void performNextWrite() {
+        ArrayList<AttributeUpdate> attributes = new ArrayList<>(2);
+        synchronized (lock) {
+            long lastEventNumber = latestEventNumbers.get(Pair.of(append.getSegment(), append.getWriterId()));
+            if (lastEventNumber == SegmentMetadata.NULL_ATTRIBUTE_VALUE) {
+                attributes.add(new AttributeUpdate(append.getWriterId(), AttributeUpdateType.None,
+                                                   append.getEventNumber()));
+            } else {
+                attributes.add(new AttributeUpdate(append.getWriterId(), AttributeUpdateType.ReplaceIfEquals,
+                                                   append.getEventNumber(), lastEventNumber));
+            }
+        }
+        attributes.add(new AttributeUpdate(EVENT_COUNT, AttributeUpdateType.Accumulate, append.getEventCount()));
+        ByteBuf buf = append.getData().asReadOnly();
         byte[] bytes = new byte[buf.readableBytes()];
         buf.readBytes(bytes);
-
-        val attributes = Arrays.asList(new AttributeUpdate(
-                        toWrite.getConnectionId(),
-                        AttributeUpdateType.ReplaceIfGreater,
-                        toWrite.getEventNumber()),
-                new AttributeUpdate(EVENT_COUNT, AttributeUpdateType.Accumulate, toWrite.getEventCount()));
-
-        CompletableFuture<Void> future;
-        String segment = toWrite.getSegment();
-        if (toWrite.isConditional()) {
-            future = store.append(segment, toWrite.getExpectedLength(), bytes, attributes, TIMEOUT);
+        if (append.isConditional()) {
+            return store.append(append.getSegment(), append.getExpectedLength(), bytes, attributes, TIMEOUT);
         } else {
-            future = store.append(segment, bytes, attributes, TIMEOUT);
+            return store.append(append.getSegment(), bytes, attributes, TIMEOUT);
         }
-        future.whenComplete((t, u) -> {
-            try {
-                boolean conditionalFailed = u != null && (ExceptionHelpers.getRealException(u) instanceof BadOffsetException);
-                synchronized (lock) {
-                    if (outstandingAppend != toWrite) {
-                        throw new IllegalStateException(
-                                "Synchronization error in: " + AppendProcessor.this.getClass().getName());
-                    }
+    }
 
-                    toWrite.getData().release();
-                    outstandingAppend = null;
-                    if (u != null && !conditionalFailed) {
-                        waitingAppends.removeAll(toWrite.getConnectionId());
-                        latestEventNumbers.remove(toWrite.getConnectionId());
-                    }
+    private void handleAppendResult(final Append append, Throwable exception) {
+        try {
+            boolean conditionalFailed = exception != null && (ExceptionHelpers.getRealException(exception) instanceof BadOffsetException);
```

Comment:
`BadOffsetException` is when you append-with-offset. You still need to check for that if you made such an append, but you need to handle `BadAttributeUpdateException` as that's what will get thrown back in case of a failed CAS.